Cython/Compiler/Optimize.py

```diff
@@ -333,6 +339,78 @@ def _transform_reversed_iteration(self, node, reversed_function):
+            slice_node.as_none_safe_node("'NoneType' is not iterable"),
+            may_hold_none=False
+            )
+
+        start_node = ExprNodes.IntNode(
+            node.pos, value='0', constant_result=0, type=PyrexTypes.c_py_ssize_t_type)
+        length_temp = UtilNodes.LetRefNode(ExprNodes.SimpleCallNode(node.pos,
+                                    function = ExprNodes.NameNode(node.pos, name="len"),
+                                    args = [unpack_temp_node]
+                                    ), type=PyrexTypes.c_py_ssize_t_type)
+        end_node = length_temp
+        if reversed:
+            relation1, relation2 = '>', '>='
+            start_node, end_node = end_node, start_node
+        else:
+            relation1, relation2 = '<=', '<'
+
+        counter_ref = UtilNodes.LetRefNode(pos=node.pos, type=PyrexTypes.c_py_ssize_t_type)
+
+        target_value = ExprNodes.IndexNode(slice_node.pos, base=unpack_temp_node,
+                                           index=counter_ref)
+
+        target_assign = Nodes.SingleAssignmentNode(
+            pos = node.target.pos,
+            lhs = node.target,
+            rhs = target_value)
+
+        # analyse with boundscheck and wraparound
+        # off (because we're confident we know the size)
+        env = self.current_env()
+        new_directives = dict(env.directives, boundscheck=False, wraparound=False)
+        new_directives.pop('test_assert_path_exists', None)
+        new_directives.pop('test_fail_if_path_exists', None)
+        target_assign = Nodes.CompilerDirectivesNode(target_assign.pos,
+                                                        directives = new_directives,
+                                                        body = target_assign)
+
+        target_assign = target_assign
+        body = Nodes.StatListNode(
+            node.pos,
+            stats = [target_assign]) # exclude node.body for now to not reanalyse it
+
+        loop_node = Nodes.ForFromStatNode(
+            node.pos,
+            bound1=start_node, relation1=relation1,
+            target=counter_ref,
+            relation2=relation2, bound2=end_node,
+            step=None, body=body,
+            else_clause=node.else_clause,
+            from_range=True)
+
+        ret = UtilNodes.LetNode(
+                    unpack_temp_node,
+                    UtilNodes.LetNode(
+                        length_temp,
+                        # we don't actually have any use for the result of "counter_ref"
+                        # but the TempResultFromStatNode UtilNode already exists and does
+                        # basically the right thing
+                        Nodes.ExprStatNode(node.pos,
+                            expr=UtilNodes.TempResultFromStatNode(
+                                    counter_ref,
+                                    loop_node
+                            )
+                        )
+                    )
+                ).analyse_expressions(env)
```

Comment:
I'm having a hard time understanding what this part does. What is "basically the right thing"?